api/src/permissions/permissions.service.spec.ts

```diff
@@ -171,7 +172,152 @@ describe("PermissionService", () => {
+                    groupId: "group-test-top",
+                    permission: [AclPermission.View],
+                },
+            ];
+            PermissionSystem.upsertGroups([groupDoc2]);
+
+            expect(
+                PermissionSystem.verifyAccess(
+                    ["group-test-middle"],
+                    DocType.Post,
+                    AclPermission.View,
+                    ["group-test-top"],
+                ),
+            ).toBe(true);
+
+            // Create a lower level group, and give group-test-middle language document edit access to it.
+            const groupDoc3: GroupDto = new GroupDto();
+            groupDoc3._id = "group-test-low";
+            groupDoc3.type = DocType.Group;
+            groupDoc3.updatedTimeUtc = 3;
+            groupDoc3.name = "Test Lower Group";
+            groupDoc3.acl = [
+                {
+                    type: DocType.Language,
+                    groupId: "group-test-middle",
+                    permission: [AclPermission.View, AclPermission.Edit],
+                },
+            ];
+            PermissionSystem.upsertGroups([groupDoc3]);
+
+            // Both group-test-top and group-test-middle should have edit access to group-test-low
+            expect(
+                PermissionSystem.verifyAccess(
+                    ["group-test-low"],
+                    DocType.Language,
+                    AclPermission.Edit,
+                    ["group-test-top"],
+                ),
+            ).toBe(true);
+            expect(
+                PermissionSystem.verifyAccess(
+                    ["group-test-low"],
+                    DocType.Language,
+                    AclPermission.Edit,
+                    ["group-test-middle"],
+                ),
+            ).toBe(true);
+
+            // Add and remove language edit access to the top level group,
+            // and check if the top level group still has language edit access to the low level group
+            const groupDoc1a = {
+                ...groupDoc1,
+                acl: [
+                    {
+                        type: DocType.Post,
+                        groupId: "group-test-top",
+                        permission: [AclPermission.View],
+                    },
+                    {
+                        type: DocType.Language,
+                        groupId: "group-test-top",
+                        permission: [AclPermission.View, AclPermission.Edit],
+                    },
+                ],
+            };
+            PermissionSystem.upsertGroups([groupDoc1a]);
+            PermissionSystem.upsertGroups([groupDoc1]);
+
+            expect(
+                PermissionSystem.verifyAccess(
+                    ["group-test-low"],
+                    DocType.Language,
+                    AclPermission.Edit,
+                    ["group-test-top"],
+                ),
+            ).toBe(true);
+
+            // Remove the test groups for further tests
```

Comment:
This feels a bit error-prone, is there no way to reset the state / create a new permission system instance for every test?